File: Source/ControllersAndRegions/LevelTimerToCounterController.cs
using Microsoft.Xna.Framework;
using Monocle;
using Celeste.Mod.Entities;
using System;

namespace Celeste.Mod.Microlith57Misc.Entities;

[CustomEntity("Microlith57Misc/LevelTimerToCounterController")]
public sealed class LevelTimerToCounterController(EntityData data, Vector2 offset) : Entity(data.Position + offset) {

    public enum WrapMode {
        NoWrapping,
        Positive,
        FullRange,
    }

    public enum PrecisionMode {
        Seconds,
        FramesProbably,
        Milliseconds,
    }

    public (WrapMode Wrap, PrecisionMode Precision) Mode
        = (data.Enum<WrapMode>("wrapMode"), data.Enum<PrecisionMode>("precisionMode"));


    public string Flag = data.Attr("flag", "");
    public bool InvertFlag = data.Bool("invertFlag");
    public string Counter = data.Attr("counter", "levelTimer");

    public override void Update() {
        base.Update();

        var session = (Scene as Level)!.Session;

        if (string.IsNullOrEmpty(Flag) || (session.GetFlag(Flag) ^ InvertFlag))
            session.SetCounter(Counter, PerformWrapping(Math.Floor(GetValue(session.Time))));
    }

    private double GetValue(long time) {
        var span = TimeSpan.FromTicks(time);

        switch (Mode.Precision) {
            case PrecisionMode.Seconds:
                return span.TotalSeconds;
            case PrecisionMode.FramesProbably:
                return span.TotalSeconds * 60f;
            case PrecisionMode.Milliseconds:
                return span.TotalMilliseconds;
            default:
                throw new Exception("should be unreachable!");
        }
    }

    private int PerformWrapping(double val) {
        if (val <= int.MaxValue)
            return (int)val;
        else switch (Mode.Wrap) {
                case WrapMode.NoWrapping:
                    return int.MaxValue;
                case WrapMode.Positive:
                    return (int)(val % ((double)int.MaxValue + 1));
                case WrapMode.FullRange:
                    return (int)val;
                default:
                    throw new Exception("should be unreachable!");
            }
    }

}

File: Loenn/entities/LevelTimerToCounterController.lua
local utils = require("utils")

local LevelTimerToCounterController = {}

LevelTimerToCounterController.name = "Microlith57Misc/LevelTimerToCounterController"
LevelTimerToCounterController.depth = -1000000
LevelTimerToCounterController.texture = "objects/microlith57/misc/level_timer_to_counter_controller"
LevelTimerToCounterController.placements = {
    {
        name = "levelTimerSecondsToCounterController",
        data = {
            flag = "",
            invertFlag = false,
            counter = "levelTimer",
            wrapMode = "NoWrapping",
            precisionMode = "Seconds"
        }
    },
    {
        name = "levelTimerFramesToCounterController",
        data = {
            flag = "",
            invertFlag = false,
            counter = "levelTimer",
            wrapMode = "NoWrapping",
            precisionMode = "FramesProbably"
        }
    },
    {
        name = "levelTimerMillisecondsToCounterController",
        data = {
            flag = "",
            invertFlag = false,
            counter = "levelTimer",
            wrapMode = "NoWrapping",
            precisionMode = "Milliseconds"
        }
    }
}
LevelTimerToCounterController.fieldInformation = {
    wrapMode = {
        options = {
            "NoWrapping",
            "Positive",
            "FullRange"
        }
    },
    precisionMode = {
        options = {
            "Seconds",
            "FramesProbably",
            "Milliseconds"
        }
    }
}

return LevelTimerToCounterController
